nilearn/mass_univariate/permuted_least_squares.py

```diff
@@ -446,92 +652,234 @@ def permuted_ols(
+                temp_labeled_arr3d, _ = ndimage.measurements.label(
+                    scores_original_data_3d < -threshold_t,
+                    conn,
+                )
+                temp_labeled_arr3d[
+                    temp_labeled_arr3d > threshold_t
+                ] += n_positive_clusters
+                labeled_arr3d = labeled_arr3d + temp_labeled_arr3d
+                del temp_labeled_arr3d
+
+            cluster_labels, idx, cluster_sizes = np.unique(
+                labeled_arr3d,
+                return_inverse=True,
+                return_counts=True,
+            )
+            assert cluster_labels[0] == 0  # the background
+
+            # Cluster mass-based inference
+            cluster_masses = np.zeros(cluster_labels.shape)
+            for j_val in cluster_labels[1:]:  # skip background
+                cluster_mass = np.sum(
+                    np.fabs(scores_original_data_3d[labeled_arr3d == j_val])
+                    - threshold_t
+                )
+                cluster_masses[j_val] = cluster_mass
+
+            p_cmfwe_vals = _null_to_p(
+                cluster_masses,
+                cmfwe_h0[i_regressor, :],
+                'upper',
+            )
+            p_cmfwe_map = p_cmfwe_vals[np.reshape(idx, labeled_arr3d.shape)]
+
+            # Convert 3D to image, then to 1D
+            # There is a problem if the masker performs preprocessing,
+            # so we use apply_mask here.
+            cmfwe_pvals[i_regressor, :] = np.squeeze(
+                apply_mask(
+                    nib.Nifti1Image(
+                        p_cmfwe_map,
+                        masker.mask_img_.affine,
+                        masker.mask_img_.header,
+                    ),
+                    masker.mask_img_,
+                )
+            )
+
+            # Cluster size-based inference
+            cluster_sizes[0] = 0  # replace background's "cluster size" w zeros
+            p_csfwe_vals = _null_to_p(
+                cluster_sizes,
+                csfwe_h0[i_regressor, :],
+                'upper',
+            )
+            p_csfwe_map = p_csfwe_vals[np.reshape(idx, labeled_arr3d.shape)]
+
+            # There is a problem if the masker performs preprocessing,
+            # so we use apply_mask here.
+            csfwe_pvals[i_regressor, :] = np.squeeze(
+                apply_mask(
+                    nib.Nifti1Image(
+                        p_csfwe_map,
+                        masker.mask_img_.affine,
+                        masker.mask_img_.header,
+                    ),
+                    masker.mask_img_,
+                )
+            )
+
+    if output_type == 'legacy':
+        outputs = (-np.log10(vfwe_pvals), scores_original_data.T, vfwe_h0)
+    else:
+        outputs = {
+            't': scores_original_data.T,
+            'logp_max_t': -np.log10(vfwe_pvals),
+            'h0_max_t': vfwe_h0,
+        }
+
+        if threshold is not None:
```

Comment:
Just wondering: shouldn't cluster sizes/mass also not be part of the output, as is `t` for voxel-level statistics ?

Reply:
Sorry, this question confused me a bit. Could you expand on it?

Reply:
Sorry,
For voxel level inference, you're returning the `t` values, together with null distribution and p-values.
For cluster level inference, why not returning the cluster size statistic (as an image), together with he null distribution of max cluster sizes and p-values ?
Or did I miss something ?

Reply:
Ah, okay. I've added that in [f1eebf5](https://github.com/nilearn/nilearn/pull/3181/commits/f1eebf514df388f3ed94a18e3be0d417bfdedc87).